src/control/common/system.go

```
@@ -0,0 +1,132 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package common
+
+import (
+	"sync"
+
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/daos-stack/daos/src/control/logging"
+	"github.com/pkg/errors"
+)
+
+// SystemMember refers to a data-plane instance that is a member of this DAOS
+// system running on host with the control-plane listening at "Addr".
+type SystemMember struct {
+	Addr string
+	Uuid string
+	Rank uint32
+}
+
+// Membership tracks details of system members.
+type Membership struct {
+	sync.Mutex
+	log     logging.Logger
+	members map[string]SystemMember
+}
+
+// Add adds member to membership.
+func (m *Membership) Add(member SystemMember) error {
+	m.Lock()
+	defer m.Unlock()
+
+	if value, found := m.members[member.Uuid]; found {
+		return errors.Errorf("member %s already exists at %+v",
+			member.Uuid, value)
+	}
+
+	m.members[member.Uuid] = member
+	m.log.Debugf("system membership: %+v\n", m.members)
+
+	return nil
+}
+
+// Remove removes member from membership.
+func (m *Membership) Remove(uuid string) error {
```

Not sure if this should really return an error, because it means the caller has to check it. Does the caller really care if the to-be-removed member doesn't exist? I think we can take a cue from the delete() built-in... It does not return an error. If the map is nil or the requested member doesn't exist, the operation is a no-op.